src/index.js

```diff
@@ -313,145 +289,174 @@ function addSegment(currentState, segment) {
+      currentState = currentState.put({ validChars: '/' });
+      regex += '/';
 
-      var handler = { handler: route.handler, names: names };
-      handlers.push(handler);
+      // Add a representation of the segment to the NFA and regex
+      currentState = addSegment(currentState, segment);
+      regex += segment.regex();
     }
 
     if (isEmpty) {
-      currentState = currentState.put({ validChars: "/" });
-      regex += "/";
+      currentState = currentState.put({ validChars: '/' });
+      regex += '/';
     }
 
-    currentState.handlers = handlers;
-    currentState.regex = new RegExp(regex + "$");
-    currentState.types = types;
+    var handlers = [{ handler: route.handler, names: names }];
 
-    if (name = options && options.as) {
-      this.names[name] = {
-        segments: allSegments,
+    if (routeName) {
+      this.names[routeName] = {
+        segments: segments,
         handlers: handlers
       };
     }
-  },
 
-  handlersFor: function(name) {
-    var route = this.names[name], result = [];
-    if (!route) { throw new Error("There is no route named " + name); }
+    currentState.handlers = handlers;
+    currentState.regex = new RegExp(regex + '$');
+    currentState.types = types;
+  }
+
+  handlersFor(name) {
+    var route = this.names[name],
+        result = [];
+
+    if (!route) {
+      throw new Error(`There is no route named ${name}`);
+    }
 
     for (var i=0, l=route.handlers.length; i<l; i++) {
       result.push(route.handlers[i]);
     }
 
     return result;
-  },
+  }
 
-  hasRoute: function(name) {
+  hasRoute(name) {
     return !!this.names[name];
-  },
+  }
+
+  generate(name, params) {
+    var route = this.names[name], output = '';
 
-  generate: function(name, params) {
-    var route = this.names[name], output = "";
-    if (!route) { throw new Error("There is no route named " + name); }
+    if (!route) {
+      throw new Error(`There is no route named ${name}`);
+    }
 
     var segments = route.segments;
 
-    for (var i=0, l=segments.length; i<l; i++) {
+    for (var i = 0, l = segments.length; i < l; i++) {
       var segment = segments[i];
 
-      if (segment instanceof EpsilonSegment) { continue; }
+      if (segment instanceof EpsilonSegment) {
+        continue;
+      }
+
+      output += '/';
+      var segmentValue = segment.generate(params);
+      if (segmentValue === undefined) {
+        throw new Error(`A value is required for route parameter '${segment.name}' in route '${name}'.`);
+      }
 
-      output += "/";
-      output += segment.generate(params);
+      output += segmentValue;
     }
 
-    if (output.charAt(0) !== '/') { output = '/' + output; }
+    if (output.charAt(0) !== '/') {
+      output = '/' + output;
+    }
 
     if (params && params.queryParams) {
       output += this.generateQueryString(params.queryParams, route.handlers);
     }
 
     return output;
-  },
+  }
+
+  generateQueryString(params) {
```

Comment:
Is it important for Aurelia to have zero external dependencies? If not it might be nice to avoid maintaining this code with a project like https://www.npmjs.com/package/querystring.